File: src/drivers/pic.cpp
#include <drivers/pic.hpp>
#include <io/serial.hpp>
#include <lib/logger.hpp>
#define PIC_ASSERT(n) logger::assert(initialized, "PIC[" n "]", "PIC is not initialized");

using namespace pic;

constexpr int PIC1 = 0x20;
constexpr int PIC2 = 0xA0;
constexpr int PIC1_DATA = PIC1 + 1;
constexpr int PIC2_DATA = PIC2 + 1;
constexpr int PIC_EOI = 0x20;

void pic::init() {
    asm volatile ("cli");
    io::out<uint8_t>(PIC1, 0x10 | 0x1);
    io::out<uint8_t>(PIC2, 0x10 | 0x1);
    io::out<uint8_t>(PIC1_DATA, 0x20);
    io::out<uint8_t>(PIC2_DATA, 0x20 + 8);
    io::out<uint8_t>(PIC1_DATA, 4);
    io::out<uint8_t>(PIC2_DATA, 2);
    io::out<uint8_t>(PIC1_DATA, 1);
    io::out<uint8_t>(PIC2_DATA, 1);
    disable();

    logger::info("PIC", "Initialized 8259 PIC");
}

void pic::eoi(uint8_t irq) {
    if (irq < 8)
        io::out<uint8_t>(PIC1, PIC_EOI);
    io::out<uint8_t>(PIC2, PIC_EOI);
}

void pic::mask(uint8_t irq) {
    uint16_t port;
    if (irq < 8)
        port = PIC1_DATA;
    else {
        port = PIC2_DATA;
        irq -= 8;
    }

    io::out<uint8_t>(port, io::in<uint8_t>(port) | (1 << irq));
}

void pic::unmask(uint8_t irq) {
    uint16_t port;
    if (irq < 8)
        port = PIC1_DATA;
    else {
        port = PIC2_DATA;
        irq -= 8;
    }

    io::out<uint8_t>(port, io::in<uint8_t>(port) & ~(1 << irq));
}

void pic::disable() {
    io::out<uint8_t>(PIC1_DATA, 0xFF);
    io::out<uint8_t>(PIC2_DATA, 0xFF);
    logger::verbose("PIC[DISABLE]", "Disabled 8259 PIC");
}

File: src/drivers/pic.zig
const log = @import("std").log.scoped(.pic);
const serial = @import("kernel").drivers.serial;

const PIC1 = 0x20;
const PIC2 = 0xA0;
const PIC1_DATA = PIC1 + 1;
const PIC2_DATA = PIC2 + 1;
const PIC_EOI = 0x20;

pub fn init() void {
    asm volatile ("cli");
    serial.out(PIC1, @as(u8, 0x10 | 0x1));
    serial.out(PIC2, @as(u8, 0x10 | 0x1));
    serial.out(PIC1_DATA, @as(u8, 0x20));
    serial.out(PIC2_DATA, @as(u8, 0x20 + 8));
    serial.out(PIC1_DATA, @as(u8, 4));
    serial.out(PIC2_DATA, @as(u8, 2));
    serial.out(PIC1_DATA, @as(u8, 1));
    serial.out(PIC2_DATA, @as(u8, 1));
    disable();
    log.info("Initialized 8259 PIC", .{});
}

pub inline fn eoi(irq: u8) void {
    serial.out(if (irq < 8) PIC1 else PIC2, @as(u8, PIC_EOI));
}

pub fn mask(_irq: u8) void {
    var irq = _irq;
    const port = get_port(&irq);
    serial.out(port, serial.in(port, u8) | (@as(u8, 1) << @intCast(irq)));
}

pub fn unmask(_irq: u8) void {
    var irq = _irq;
    const port = get_port(&irq);
    serial.out(port, serial.in(port, u8) & ~(@as(u8, 1) << @intCast(irq)));
}

fn get_port(irq: *u8) u16 {
    if (irq.* < 8) return PIC1_DATA;
    irq.* -= 8;
    return PIC2_DATA;
}

pub fn disable() void {
    serial.out(PIC1_DATA, @as(u8, 0xFF));
    serial.out(PIC2_DATA, @as(u8, 0xFF));
}
